pkg/operator/ceph/pool/controller.go

```diff
@@ -18,155 +18,185 @@ limitations under the License.
+func Add(mgr manager.Manager, context *clusterd.Context) error {
+	return add(mgr, newReconciler(mgr, context))
 }
 
-// Watch watches for instances of Pool custom resources and acts on them
-func (c *PoolController) StartWatch(namespace string, stopCh chan struct{}) {
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, context *clusterd.Context) reconcile.Reconciler {
+	// Add the cephv1 scheme to the manager scheme so that the controller knows about it
+	mgrScheme := mgr.GetScheme()
+	cephv1.AddToScheme(mgr.GetScheme())
 
-	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
-		AddFunc:    c.onAdd,
-		UpdateFunc: c.onUpdate,
-		DeleteFunc: c.onDelete,
+	return &ReconcileCephBlockPool{
+		client:  mgr.GetClient(),
+		scheme:  mgrScheme,
+		context: context,
 	}
-
-	logger.Infof("start watching pools in namespace %q", namespace)
-	go k8sutil.WatchCR(PoolResource, namespace, resourceHandlerFuncs, c.context.RookClientset.CephV1().RESTClient(), &cephv1.CephBlockPool{}, stopCh)
 }
 
-func (c *PoolController) onAdd(obj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Creating pools for an external ceph cluster is disabled because no Ceph image is specified")
-		return
-	}
-
-	pool, err := getPoolObject(obj)
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		logger.Errorf("failed to get pool object. %v", err)
-		return
+		return err
 	}
-	updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ProcessingStatus, c.context)
-	err = createPool(c.context, pool)
+
+	// Watch for changes on the CephBlockPool CRD object
+	err = c.Watch(&source.Kind{Type: &cephv1.CephBlockPool{}}, &handler.EnqueueRequestForObject{})
 	if err != nil {
-		logger.Errorf("failed to create pool %q. %v", pool.ObjectMeta.Name, err)
-		updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.FailedStatus, c.context)
-		return
+		return err
 	}
-	updateCephBlockPoolStatus(pool.GetName(), pool.GetNamespace(), k8sutil.ReadyStatus, c.context)
+
+	return nil
 }
 
-func (c *PoolController) onUpdate(oldObj, newObj interface{}) {
-	if c.clusterSpec.External.Enable && c.clusterSpec.CephVersion.Image == "" {
-		logger.Warningf("Updating pools for an external ceph cluster is disabled because no Ceph image is specified")
-		return
-	}
+// Reconcile reads that state of the cluster for a CephBlockPool object and makes changes based on the state read
+// and what is in the CephBlockPool.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileCephBlockPool) Reconcile(request reconcile.Request) (reconcile.Result, error) {
```

Comment:
For the other controllers, I've wrapped the reconciler to log returned errors[0], but I see you've logged wherever you're returning. The other way would be more consistent with controller-runtime practices (to return the error and have it automatically logged). The only reason I had to wrap[0] is because controller-runtime logging interface is different.

This works too, but it might not occur to people coming from controller-runtime controllers that returned errors aren't automatically logged.

Not a big deal though.

[0] https://github.com/rook/rook/blob/master/pkg/operator/ceph/disruption/nodedrain/reconcile.go#L63

Reply:
Will implement this, thanks